### PR TITLE
Add support for Windows

### PR DESCRIPTION
**Pull request by nehaljwani**
Bonus: Appveyor config

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -27,12 +27,22 @@
...
+           for e in self.extensions:
+               e.extra_compile_args = ["-mstackrealign"]
+        build_ext.build_extensions(self)
+
 packages = ["backports", "backports.lzma"]
 prefix = sys.prefix
 home = os.path.expanduser("~")
-extens = [Extension('backports/lzma/_lzma',
+extens = [Extension('backports.lzma._lzma',
                     ['backports/lzma/_lzmamodule.c'],
```
Comment (jakirkham) [inline setup.py]: Would have figured this would need to change too.

**Comment by jakirkham** [comment]
Thanks for doing this, @nehaljwani!

File: setup.py (hunk 1 above)
Comment (nehaljwani) [reply setup.py]: I don't follow. Is this wrong?

Comment (jakirkham) [inline setup.py]: Would have figured you would need `\` on Windows. Maybe using `os.path.join` would work?

Comment (nehaljwani) [reply setup.py]: According to the documentation of distutils, 'name' is:
```
    the full name of the extension, including any packages - ie. not a
    filename or pathname, but Python dotted name
```

Comment (nehaljwani) [reply setup.py]: Oh, you were referring to the source argument, no, that's not really required, I think. Docs say:
```
list of source filenames, relative to the distribution root (where the setup script lives), in Unix form (slash- separated) for portability. 
```

Comment (nehaljwani) [reply setup.py]: The Appveyor CI ran the build with this PR. See https://ci.appveyor.com/project/peterjc/backports-lzma/build/1.0.23

**Comment by jakirkham** [comment]
If it works, that’s fine. Just a little surprised it worked was all.

**Comment by jakirkham** [comment]
Tried using this patch in PR ( https://github.com/conda-forge/backports.lzma-feedstock/pull/4 ). However ran into an issue building on Python 2.7 64-bit, which your PR doesn't seem to run into. According to this [SO answer]( https://stackoverflow.com/a/19867426 ), we need to add `-D MS_WIN64` to the compiler arguments. Any thoughts on adding this in `setup.py` somehow?

File: test/test_lzma.py [hunk 2]
```diff
@@ -8,8 +8,12 @@
...
     from test.support import _4G, TESTFN, import_module, bigmemtest, run_unittest, unlink
 except ImportError:
-    # Must be under Python 2 then,
-    from test.test_support import _4G, TESTFN, import_module, bigmemtest, run_unittest, unlink
+    try:
+        # May be < Python 2.7.14
+        from test.test_support import _4G, TESTFN, import_module, bigmemtest, run_unittest, unlink
+    except ImportError:
+        # May be >=Python 2.7.14, <3
+        from future.backports.test.support import _4G, TESTFN, import_module, bigmemtest, run_unittest, unlink
```
Comment (jakirkham) [inline test/test_lzma.py]: With Python 2.7.14, this fallback should be unnecessary. Though `defaults` will need to rebuild with a patch. Just wrote it for `conda-forge`. Please see PR ( https://github.com/conda-forge/python-feedstock/pull/168 ) for details.

**Comment by nehaljwani** [comment]
Python from defaults has a patched distutils to take care of `-D MS_WIN64` automatically for us. See https://github.com/AnacondaRecipes/python-feedstock/blob/master-2.7.14/recipe/0008-Win32-Fixes-for-Windows-GCC-interop-needed-by-RPy2-a.patch

File: test/test_lzma.py (hunk 2 above)
Comment (nehaljwani) [reply test/test_lzma.py]: Well, for some reason the python 2.7.14 in Fedora 27 also doesn't have this, neither does the (library) docker container for python 2.7.14. The only distro where I found this working was in ArchLinux.

**Comment by nehaljwani** [comment]
I have updated the PR so that `-DMS_WIN{32,64}` are passed explicitly when the compiler is set to mingw32.

File: test/test_lzma.py (hunk 2 above)
Comment (nehaljwani) [reply test/test_lzma.py]: I reported it here: https://bugzilla.redhat.com/show_bug.cgi?id=1528899

Comment (jakirkham) [reply test/test_lzma.py]: Yeah, I think a lot of people were caught by surprise. It's not really a breaking change, but still it requires some changes for those of us aggressively cleaning `test`. Thanks for raising the issue.

**Comment by jakirkham** [comment]
Happy to patch `conda-forge`'s `python` the same way. Is there a CPython issue on this point? In any event, we should still set this for everyone else. Thanks for doing that @nehaljwani.

File: test/test_lzma.py (hunk 2 above)
Comment (jakirkham) [reply test/test_lzma.py]: Should add that the fallback is harmless (and helpful in some cases). So it is reasonable to keep.

**Comment by mingwandroid** [comment]
> Thanks for doing that @nehaljwani.

Err, I did it.

**Comment by mingwandroid** [comment]
CPython have no interest in gcc on Windows. They will repeat the mantra that 'distutils is frozen'. Many of us spent years submitting patches to them only to face a wall of disinterest and silence.

Of course you are welcome to try submitting bugs and patches again.

**Comment by jakirkham** [comment]
Was referring to commit ( https://github.com/peterjc/backports.lzma/commit/37976356226e3ec5f925bc13461e2098fd2f1dfb ). Though my comment was a bit convoluted. That said, absolutely thanks for the `python` patch, @mingwandroid.

File: setup.py (hunk 1 above)
Comment (peterjc) [reply setup.py]: Does it also work with the original Linux style slashes?

**Comment by peterjc** [comment]
You've gone a lot further than I did with https://github.com/peterjc/backports.lzma/blob/appveyor/.appveyor.yml - I think I would never have managed this on my own without setting up a replacement Windows development machine.

The fact that on top of @nehaljwani's great work we have knowledgeable review and input from @jakirkham and @mingwandroid is very reassuring.

And good news that it has been shown to work by the conda-forge team too, https://github.com/conda-forge/backports.lzma-feedstock/issues/1 - I wasn't aware it in Conda yet, so that was another surprise Christmas present.

**Comment by peterjc** [comment]
Still todo - a version bump and release, and then uploading Windows wheels to PyPI.

**Comment by peterjc** [comment]
@nehaljwani may I name you as a contributor in the README file? Should I use "Nehal J Wani" as per your GitHub profile?

**Comment by jakirkham** [comment]
Thanks @peterjc. Agree that a release would be good.

File: setup.py (hunk 1 above)
Comment (nehaljwani) [reply setup.py]: Yes, it is portable.

**Comment by nehaljwani** [comment]
@peterjc Yes, you can use the name I have on my Github Profile.

**Pull request merged**